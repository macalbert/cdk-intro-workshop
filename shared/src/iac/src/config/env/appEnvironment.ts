export enum AppEnvironment {
    Production = "Production",
    Development = "Development",
    Staging = "Staging",
    GithubAction = "GithubAction",
    Test = "Test",
}
