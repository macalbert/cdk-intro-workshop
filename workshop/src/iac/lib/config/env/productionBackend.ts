import {
    ApiGatewayStack,
    type ApiGatewayStackProps,
    AppEnvironment,
    ApiLambdaStack,
    type ApiLambdaStackProps,
    S3Stack,
    type S3StackProps,
    StackBuildPart,
    type StackBuildProps,
} from "@m47/shared-iac";
import { type Stack } from "aws-cdk-lib";
import { SubnetType } from "aws-cdk-lib/aws-ec2";
import type { IFunction } from "aws-cdk-lib/aws-lambda";

export interface ProductionBackendProps extends StackBuildProps {
    clusterName: string;
    serviceName: string;
    exposedPort: number;
    sourceCodePath: string;
    dockerfileApi: string;
}

export class ProductionBackend extends StackBuildPart {
    props: ProductionBackendProps;

    environment = AppEnvironment.Production;

    constructor(buildProps: ProductionBackendProps) {
        super(buildProps);

        this.props = buildProps;
        this.props.githubRepo = this.formatRepoNameForCloudFormation();
    }

    build(): Stack[] {
        
        const apiLambdaStack = this.createApiLambdaStack(
            "ApiLambda",
            this.props.dockerfileApi,
            "api",
        );

        const apiGatewayStack = this.createApiGateway(
            apiLambdaStack.lambdaFunction,
            "ApiGateway",
            "api-cdk-workshop",
        );

        return [
            apiLambdaStack,
            apiGatewayStack,
        ];
    }

    private createApiLambdaStack(
        name: string,
        dockerfile: string,
        subdomain: string,
    ): ApiLambdaStack {
        const lambdaExportProps: ApiLambdaStackProps = {
            name: name,
            githubRepo: this.props.githubRepo,
            pathDockerFile: this.props.sourceCodePath,
            env: this.props.env,
            vpc: this.props.vpc,
            envName: AppEnvironment.Production,
            timeoutSeconds: 29,
            memorySizeMbs: 1024,
            vpcSubnets: {
                subnetType: SubnetType.PRIVATE_WITH_EGRESS,
            },
            stackName: `${this.props.githubRepo}-${name}`,
            functionName: `${this.props.githubRepo}-${subdomain}`.toLowerCase(),
            dockerFile: dockerfile,
        };

        const stack = new ApiLambdaStack(this.props.scope, lambdaExportProps);

        return stack;
    }

    private createApiGateway(
        lambdaFunction: IFunction,
        name: string,
        subdomain: string,
    ): ApiGatewayStack {
        const apiProps: ApiGatewayStackProps = {
            env: this.props.env,
            name: name,
            envName: AppEnvironment.Production,
            lambdaFunction: lambdaFunction,
            stackName: `${this.props.githubRepo}-${name}`,
            subdomain: subdomain,
            githubRepo: this.props.githubRepo,
            certificateArn: `arn:aws:acm:us-east-1:${this.props.env.account}:certificate/be63062d-5316-47af-9f94-819c1dc02853`,
            domain: "m47.io",
        };

        return new ApiGatewayStack(this.props.scope, apiProps);
    }
}
