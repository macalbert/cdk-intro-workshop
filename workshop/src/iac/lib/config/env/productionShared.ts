import {
    AppEnvironment,
    PipelineStack,
    type PipelineStackProps,
    RdsStack,
    type RdsStackProps,
    StackBuildPart,
    type StackBuildProps,
    WindowsBastionStack,
    type WindowsBastionStackProps,
} from "@m47/shared-iac";
import { RemovalPolicy, type Stack } from "aws-cdk-lib";
import { InstanceType, InstanceClass, InstanceSize } from "aws-cdk-lib/aws-ec2";
import { PostgresEngineVersion } from "aws-cdk-lib/aws-rds";

export class ProductionShared extends StackBuildPart {
    props: StackBuildProps;

    constructor(buildProps: StackBuildProps) {
        super(buildProps);

        this.props = buildProps;
    }

    build(): Stack[] {
        const pipelineStack = this.createPipelineStack();
        const rdsStack = this.createRdsStack();
        const createBastionStack = this.createBastionStack();

        return [
            rdsStack,
            pipelineStack,
            createBastionStack,
        ];
    }

    private createPipelineStack(): PipelineStack {
        const pipeline: PipelineStackProps = {
            name: "Pipeline",
            branch: this.props.branch,
            githubRepo: this.props.githubRepo,
            envName: AppEnvironment.Production,
            secretTokenArn: `arn:aws:secretsmanager:${this.props.env.region}:${this.props.env.account}:secret:github-access-token-secret-dth2kY`,
            testBuildSpec: [
                "workshop/src/iac/buildspecs/test/iac.yml",
                "workshop/src/iac/buildspecs/test/backend.yml",
            ],
            deployBuildSpec: [
                "workshop/src/iac/buildspecs/production/backendApi.yml",
            ],
            testProjectName: `${this.props.githubRepo}-Tests`,
            vpc: this.props.vpc,
            env: this.props.env,
            bucketRemovalPolicy: RemovalPolicy.DESTROY,
            domain: "workshop.com",
            stackName: `${this.formatRepoNameForCloudFormation()}-CodePipeline`,
            manualApproval: false,
            shouldCreateNewSlackChannel: false,
        };

        return new PipelineStack(this.props.scope, pipeline);
    }

    private createRdsStack(): RdsStack {
        const props: RdsStackProps = {
            env: this.props.env,
            vpc: this.props.vpc,
            databaseName: "workshop",
            databaseUserName: "masteruser",
            backupRetentionDays: 14,
            storageSizeGb: 100,
            monitoringIntervalSeconds: 60,
            engineVersion: PostgresEngineVersion.VER_17_4,
            instanceType: InstanceType.of(InstanceClass.T3, InstanceSize.MICRO),
            githubRepo: this.props.githubRepo,
            envName: AppEnvironment.Production,
            stackName: `${this.formatRepoNameForCloudFormation()}-rds`,
            name: "RdsPostgres",
        };

        return new RdsStack(this.props.scope, props);
    }

    private createBastionStack(): WindowsBastionStack {
        const windowsBastionStackProps: WindowsBastionStackProps = {
            vpc: this.props.vpc,
            githubRepo: this.props.githubRepo,
            envName: AppEnvironment.Production,
            name: "bastion",
            stackName: `${this.formatRepoNameForCloudFormation()}-Bastion`,
            env: this.props.env,
        };

        return new WindowsBastionStack(
            this.props.scope,
            windowsBastionStackProps,
        );
    }
}
