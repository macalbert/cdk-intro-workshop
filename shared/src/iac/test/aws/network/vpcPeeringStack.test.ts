import { App, Stack } from "aws-cdk-lib";
import { Template } from "aws-cdk-lib/assertions";
import {
    VpcPeeringStack,
    type VpcPeeringStackProps,
    type VpcPeer,
} from "../../../src/aws/network/vpcPeeringStack";
import { AppEnvironment } from "../../../src/config/env/appEnvironment";
import { Vpc } from "aws-cdk-lib/aws-ec2";

describe("VpcPeeringStack", () => {
    const env = {
        account: "account",
        region: "eu-west-1",
    };

    beforeAll(() => {
        jest.spyOn(Vpc, "fromLookup").mockImplementation(
            (scope, id, options) => {
                return {
                    vpcId: options.vpcId,
                    publicSubnets: [
                        { routeTable: { routeTableId: "rtb-public-1" } },
                        { routeTable: { routeTableId: "rtb-public-2" } },
                    ],
                    privateSubnets: [
                        { routeTable: { routeTableId: "rtb-private-1" } },
                        { routeTable: { routeTableId: "rtb-private-2" } },
                    ],
                } as Vpc;
            },
        );
    });
    afterAll(() => {
        jest.restoreAllMocks();
    });

    test("Should_CreateVpcPeeringConnection_When_StackIsCalled", () => {
        // Arrange
        const stack = new Stack(new App(), "VpcPeeringStackTest", { env });

        const props: VpcPeeringStackProps = {
            name: "VpcPeeringStack",
            stackName: "shared",
            envName: AppEnvironment.Test,
            githubRepo: "shared-test",
            source: {
                vpcId: "vpc-0dddac3bbc04c06f9",
                destinationCidrBlock: "172.31.0.0/16",
            } as VpcPeer,
            destination: {
                vpcId: "vpc-0993f29cd9a32fcbb",
                destinationCidrBlock: "10.0.0.0/16",
            } as VpcPeer,
        };

        // Act
        const actual = new VpcPeeringStack(stack, props);

        // Assert
        const template = Template.fromStack(actual).toJSON();
        expect(template).toMatchSnapshot("VpcPeeringStackTest");
    });
});
