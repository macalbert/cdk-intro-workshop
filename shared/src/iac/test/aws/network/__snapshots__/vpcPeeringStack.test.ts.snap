// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VpcPeeringStack Should_CreateVpcPeeringConnection_When_StackIsCalled: VpcPeeringStackTest 1`] = `
{
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "DestinationToSourceRoutePrivate0": {
      "Properties": {
        "DestinationCidrBlock": "172.31.0.0/16",
        "RouteTableId": "rtb-private-1",
        "VpcPeeringConnectionId": {
          "Ref": "VpcPeering",
        },
      },
      "Type": "AWS::EC2::Route",
    },
    "DestinationToSourceRoutePrivate1": {
      "Properties": {
        "DestinationCidrBlock": "172.31.0.0/16",
        "RouteTableId": "rtb-private-2",
        "VpcPeeringConnectionId": {
          "Ref": "VpcPeering",
        },
      },
      "Type": "AWS::EC2::Route",
    },
    "DestinationToSourceRoutePublic0": {
      "Properties": {
        "DestinationCidrBlock": "172.31.0.0/16",
        "RouteTableId": "rtb-public-1",
        "VpcPeeringConnectionId": {
          "Ref": "VpcPeering",
        },
      },
      "Type": "AWS::EC2::Route",
    },
    "DestinationToSourceRoutePublic1": {
      "Properties": {
        "DestinationCidrBlock": "172.31.0.0/16",
        "RouteTableId": "rtb-public-2",
        "VpcPeeringConnectionId": {
          "Ref": "VpcPeering",
        },
      },
      "Type": "AWS::EC2::Route",
    },
    "SourceToDestinationRoutePrivate0": {
      "Properties": {
        "DestinationCidrBlock": "10.0.0.0/16",
        "RouteTableId": "rtb-private-1",
        "VpcPeeringConnectionId": {
          "Ref": "VpcPeering",
        },
      },
      "Type": "AWS::EC2::Route",
    },
    "SourceToDestinationRoutePrivate1": {
      "Properties": {
        "DestinationCidrBlock": "10.0.0.0/16",
        "RouteTableId": "rtb-private-2",
        "VpcPeeringConnectionId": {
          "Ref": "VpcPeering",
        },
      },
      "Type": "AWS::EC2::Route",
    },
    "SourceToDestinationRoutePublic0": {
      "Properties": {
        "DestinationCidrBlock": "10.0.0.0/16",
        "RouteTableId": "rtb-public-1",
        "VpcPeeringConnectionId": {
          "Ref": "VpcPeering",
        },
      },
      "Type": "AWS::EC2::Route",
    },
    "SourceToDestinationRoutePublic1": {
      "Properties": {
        "DestinationCidrBlock": "10.0.0.0/16",
        "RouteTableId": "rtb-public-2",
        "VpcPeeringConnectionId": {
          "Ref": "VpcPeering",
        },
      },
      "Type": "AWS::EC2::Route",
    },
    "VpcPeering": {
      "Properties": {
        "PeerRegion": {
          "Ref": "AWS::Region",
        },
        "PeerVpcId": "vpc-0993f29cd9a32fcbb",
        "VpcId": "vpc-0dddac3bbc04c06f9",
      },
      "Type": "AWS::EC2::VPCPeeringConnection",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
